pkg/xds/envoy/listeners/v3/dns_configurer.go

```diff
@@ -0,0 +1,99 @@
+package v3
+
+import (
+	"sort"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_data_dns "github.com/envoyproxy/go-control-plane/envoy/data/dns/v3"
+	envoy_dns "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/udp/dns_filter/v3alpha"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"github.com/kumahq/kuma/pkg/util/proto"
+)
+
+type DNSConfigurer struct {
+	VIPs         map[string]string
+	EmptyDNSPort uint32
+}
+
+func (c *DNSConfigurer) Configure(listener *envoy_listener.Listener) error {
+	pbst, err := proto.MarshalAnyDeterministic(c.dnsFilter())
+	if err != nil {
+		return err
+	}
+
+	listener.ListenerFilters = append(listener.ListenerFilters, &envoy_listener.ListenerFilter{
+		Name: "envoy.filters.udp.dns_filter",
+		ConfigType: &envoy_listener.ListenerFilter_TypedConfig{
+			TypedConfig: pbst,
+		},
+	})
+	return nil
+}
+
+func (c *DNSConfigurer) dnsFilter() *envoy_dns.DnsFilterConfig {
+	var virtualDomains []*envoy_data_dns.DnsTable_DnsVirtualDomain
+	for domain, ip := range c.VIPs {
+		virtualDomains = append(virtualDomains, &envoy_data_dns.DnsTable_DnsVirtualDomain{
+			Name: domain,
+			Endpoint: &envoy_data_dns.DnsTable_DnsEndpoint{
+				EndpointConfig: &envoy_data_dns.DnsTable_DnsEndpoint_AddressList{
+					AddressList: &envoy_data_dns.DnsTable_AddressList{
+						Address: []string{
+							ip,
+						},
+					},
+				},
+			},
+			AnswerTtl: nil,
+		})
+	}
+	sort.Stable(DnsTableByName(virtualDomains)) // for stable Envoy config
+
+	return &envoy_dns.DnsFilterConfig{
+		StatPrefix: "kuma_dns",
+		ClientConfig: &envoy_dns.DnsFilterConfig_ClientContextConfig{
+			// We configure upstream resolver to resolver that always returns that it could not find the domain (NXDOMAIN)
+			// As for this moment there is no setting to disable upstream resolving.
+			UpstreamResolvers: []*envoy_core.Address{
+				{
+					Address: &envoy_core.Address_SocketAddress{
+						SocketAddress: &envoy_core.SocketAddress{
+							Address: "127.0.0.1",
+							PortSpecifier: &envoy_core.SocketAddress_PortValue{
+								PortValue: c.EmptyDNSPort,
+							},
+						},
+					},
+				},
+			},
+		},
+		ServerConfig: &envoy_dns.DnsFilterConfig_ServerContextConfig{
+			ConfigSource: &envoy_dns.DnsFilterConfig_ServerContextConfig_InlineDnsTable{
+				InlineDnsTable: &envoy_data_dns.DnsTable{
+					VirtualDomains:     virtualDomains,
+					ExternalRetryCount: 0,
+					KnownSuffixes: []*v3.StringMatcher{
+						{
+							MatchPattern: &v3.StringMatcher_SafeRegex{
+								SafeRegex: &v3.RegexMatcher{
+									EngineType: &v3.RegexMatcher_GoogleRe2{},
+									Regex:      ".*",
```

Comment:
hmm, shall we filter by suffix here? Maybe we can avoid some of the lookups to the EmptyDNS?

Reply:
The thing is that it will not avoid upstream lookups. It will avoid trying to resolve it manually

Reply:
OK, lets go with this and see how it behaves.
Can you please add a small TODO here, to make this standing out when viewing to code. So we can check it if needed later.